python/taichi/lang/matrix.py

```diff
@@ -539,51 +529,47 @@ def make_grad(self):
             ret.entries[i] = self.entries[i].grad
         return ret
 
-    @taichi_scope
     def sum(self):
         ret = self.entries[0]
         for i in range(1, len(self.entries)):
             ret = ret + self.entries[i]
         return ret
 
-    @taichi_scope
     def norm(self, l=2, eps=0):
+        import taichi as ti
         assert l == 2
-        return impl.sqrt(self.norm_sqr() + eps)
+        return ti.sqrt(self.norm_sqr() + eps)
 
-    @taichi_scope
     def norm_sqr(self):
         return (self**2).sum()
 
-    @taichi_scope
     def max(self):
+        import taichi as ti
         ret = self.entries[0]
         for i in range(1, len(self.entries)):
-            ret = impl.max(ret, self.entries[i])
+            ret = ti.max(ret, self.entries[i])
         return ret
 
-    @taichi_scope
     def min(self):
+        import taichi as ti
         ret = self.entries[0]
         for i in range(1, len(self.entries)):
-            ret = impl.min(ret, self.entries[i])
+            ret = ti.min(ret, self.entries[i])
         return ret
 
-    @taichi_scope
     def any(self):
         import taichi as ti
-        ret = (self.entries[0] != ti.expr_init(0))
+        ret = ti.cmp_ne(self.entries[0], 0)
         for i in range(1, len(self.entries)):
-            ret = ret + (self.entries[i] != ti.expr_init(0))
-        return -(ret < ti.expr_init(0))
+            ret = ret + ti.cmp_ne(self.entries[i], 0)
+        return ti.cmp_lt(ret, 0)
 
-    @taichi_scope
     def all(self):
         import taichi as ti
-        ret = self.entries[0] != ti.expr_init(0)
+        ret = ti.cmp_ne(self.entries[0], 0)
         for i in range(1, len(self.entries)):
-            ret = ret + (self.entries[i] != ti.expr_init(0))
-        return -(ret == ti.expr_init(-len(self.entries)))
+            ret = ret + ti.cmp_ne(self.entries[i], 0)
+        return ti.cmp_eq(ret, -len(self.entries))
```

Comment:
Does it return -1 before this PR but return 1 after this PR when the result is `True`?

Reply:
OIC, should we make expr_cmp_eq return -1 one True all the time?

Reply:
We should return 1 on True after we have `ti.u1`. For now, I think we return -1 for efficiency reasons, as we can optimize `-1 and a` to `a`.

Reply:
Let's simply don't change the behavior of this function in this PR. Although we're not consistent on the value of True now, this PR already contains a large changeset.